### PR TITLE
Add a shortcut that brings the cursor to the search input field

### PR DESCRIPTION
**Pull request by MAX-786**
### Plone Documentation version affected
- [x] 6.x

## Feature
When the `ctrl+k` shortcut is pressed, the cursor is automatically placed in the search input field.

## Additional Changes
I am considering adding the `^k` to the right side of the search input field OR changing the placeholder to `use "ctrl+k" to set focus`. Before I implement this feature, I would like to get feedback from others. Please let me know your thoughts.

### CODE REVIEW

File: docs/_templates/page.html
Comment (stevepiercy) [inline docs/_templates/page.html]: - Consider macOS, not just Windows and Linux.
- Let's use readable names of keys, instead of obtuse ASCII codes.
- [`which` is deprecated](https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/which).

```suggestion
            if ((event.ctrlKey || event.metaKey) && event.key == "k") {
```

The MDN docs are awesome for testing and referring to this stuff. Random sources are not your friend.

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key#result

Comment (stevepiercy) [reply docs/_templates/page.html]: We need to give users some visual indication that a keyboard shortcut even exists. There are two possibilities I can think of, but maybe @ksuess has better ideas.

1. Modify the `placeholder` attribute on both the sidebar Search box and Search page.
2. Add a `title` attribute.
3. Something else?

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Sorry for the delay in responding. I missed that this PR was open.

**Review by MAX-786 — COMMENTED** [review]
Thanks for pointing this out to me 😊.

File: docs/_templates/page.html
Comment (MAX-786) [reply docs/_templates/page.html]: I also think that we need some indication, so `placeholder` might be a good idea, for now, ig. and for another PR I have suggested this: 
> I am considering adding the `^k` to the right side of the search input field OR changing the placeholder to `use "ctrl+k" to set focus`. Before I implement this feature, I would like to get feedback from others. Please let me know your thoughts.

Screenshot_20230102_125451

Comment (stevepiercy) [reply docs/_templates/page.html]: So far this is good.

However, we should strive to fully implement new features. Whether the display of the shortcut is part of this PR or another is up to you. I cannot merge this PR by itself because it needs to have the other part to be complete.

After thinking it through and following the inspiration of the [Crema docs](https://docs.cremawork.com/overview/folder-structure), here is my suggestion. The screenshot is on desktop.

Screen Shot 2023-01-02 at 12 58 13 AM

- Insert a `span` with the HTML markup below for the sidebar search, but move the CSS from inline to our `custom.css`. I chose the colors based on the Plone palette and for sufficient contrast.
- Add a JavaScript that detects the platform and adapts the keyboard shortcut accordingly. For macOS and iOS, use `⌘K`, and for all others use `^K`.
- Apply this feature to both the sidebar search and Search page.
- On the Search page only, hide or remove the sidebar search input.
- When a user starts typing in the search input, hide the `span`. This will avoid a conflict with the clear icon of the input of `type="search"`.
- Adapt the CSS for both mobile and desktop through [media queries](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries). The `font-size`, `padding`, and `position` attributes might need to be adjusted.
- On the search page, use `id="search-shortcut-page"`, to make it easy for CSS selectors.

```html
<span id="search-shortcut-sidebar" style="
    position: absolute;
    right: 20px;
    border-radius: 0.25em;
    background: #d9eef9;
    padding: 0.2em;
    color: #4a4a4a;
">⌘K</span>
```

Comment (MAX-786) [reply docs/_templates/page.html]: >  However, we should strive to fully implement new features. Whether the display of the shortcut is part of this PR or another is up to you. I cannot merge this PR by itself because it needs to have the other part to be complete.

Oh, No problem. Sure, then I'll make this PR complete by adding a display of the shortcut. I'll go through all these inputs you have given and I'll let you know when this PR is ready for review.
Thanks😊!

Comment (MAX-786) [reply docs/_templates/page.html]: > Insert a span with the HTML markup below for the sidebar search, but move the CSS from inline to our custom.css. I chose the colors based on the Plone palette and for sufficient contrast.

Hello, @stevepiercy I am sorry for the late response, but I'm trying to add a span element to [`search-field.html`](https://github.dev/pydata/pydata-sphinx-theme/blob/main/src/pydata_sphinx_theme/theme/pydata_sphinx_theme/components/search-field.html), but I'm having trouble finding it. I've looked in the usual places ( read [docs](https://sphinx-book-theme.readthedocs.io/en/stable/customize/sidebar-primary.html?highlight=search#default-sidebar-elements) already), but I'm unsure where else to look. Could you please help me locate it?

I can do that with JS but it may cause the UX problem IG. 
Does this looks good except for visual indication:


https://user-images.githubusercontent.com/99530996/210418460-d43f2965-b023-40f2-aba9-571e6b178ec1.mp4

Comment (stevepiercy) [reply docs/_templates/page.html]: This is a Sphinx thing. A great overview of how Sphinx themes work is at https://www.sphinx-doc.org/en/master/templating.html.

You need to override the ancestral theme files, which are located in the Python virtual environment. Do not edit files in the virtual environment, but instead place a file with the same name and relative location as the ancestral themes in our local theme overrides in `docs/_templates`. For example, in `docs/_templates/layout.html`, `{% extends "sphinx_book_theme/layout.html" %}` means extend the parent theme file `layout.html` with the following stuff.

That should help you get started, but here are few more points.

`lib/python3.10/site-packages/sphinx_book_theme/theme/sphinx_book_theme/sections/sidebar.html` says `{%- for sidebartemplate in sidebars %}`.

In its config file `lib/python3.10/site-packages/sphinx_book_theme/theme/sphinx_book_theme/theme.conf`, an array of `sidebars` is defined as `sidebars = sidebar-logo.html, search-field.html, sbt-sidebar-nav.html`. If the file does not exist in the parent theme, go up a generation to the grandparent theme `lib/python3.10/site-packages/pydata_sphinx_theme/theme/pydata_sphinx_theme/search-field.html`.

The great-grandparent theme is `lib/python3.10/site-packages/sphinx/themes/basic` by virtue of `{%- extends "basic/layout.html" %}`.

As you can see, it takes some time to learn this stuff, but the good thing is that theming in Volto is very similar to that in Sphinx. It also often similar to other theming engines, so it is a good skill to acquire, as the experience is transferrable.

Comment (MAX-786) [reply docs/_templates/page.html]: > This is a Sphinx thing. A great overview of how Sphinx themes work is at https://www.sphinx-doc.org/en/master/templating.html.

Oh, I didn't look into templating section 😅.

I may require a bit of time (also since my semester break is over😖) to go through all this but I'll make sure to do it in a day or 2, because theming in Volto is also what I am interested in.

Thanks a lot for taking the time to help me with this issue. Your guidance is really helpful and I really appreciate it and its really means a lot to me (as a newbie).

**Comment by MAX-786** [comment]
Done!

**Comment by stevepiercy** [comment]
This is looking really good.

You won't believe this, but [sphinx-book-theme just cut a new release candidate version 0.4.0rc1](https://github.com/executablebooks/sphinx-book-theme/pull/666) that partially includes this feature. I think our implementation is better. :grin:

Ours:
https://deploy-preview-1405--6-dev-docs-plone-org.netlify.app/search.html?q=hello

Theirs:
https://sphinx-book-theme.readthedocs.io/en/latest/changelog.html

No matter what they do, I think that we should keep what we have, and demo it to the parent theme maintainers for their consideration to adopt. If they like it, we can submit a PR to make it happen.

Here are few more corrections for this PR and one final comment. I would like @ksuess to weigh in, as she has good insight on design and usability.

1. On the Search page, the keyboard shortcut needs to be moved a little to the left, so that it does not touch the "search" button.
Screen Shot 2023-01-05 at 4 09 02 AM
2. On the Search page, the input is of `type="text"`, but should be `type="search"` to have the clear "X" indicator show up when someone starts typing, as in the sidebar.
3. In the sidebar, the placeholder text "Search the docs ..." should have the periods replaced with an ellipsis to be consistent with the placeholder on the Search page, "Search the docs…".
4. After approval, but before merge, I would like the inline JavaScript moved to an external file that gets added via the Sphinx configuration: https://github.com/plone/documentation/blob/24d2f6cef52fd519e32353b11d0bfd27abcac838/docs/conf.py#L131-L133. Details of how it works: https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_js_files. It's OK to inline it while developing to avoid caching issues.
5. For now, I am OK with making the search form not visible in the sidebar on the Search page, even though it takes up space. I think it might be jarring to the user if the navigation jumps up and down between searching and not searching. Navigation should not move around. However, in the long term, I think I would prefer what the grandparent theme [pydata-sphinx-theme](https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html) does for its search, specifically moving it into an icon in the top navbar and using a modal. This would require more discussion outside of this PR, and I am only mentioning it for future discussion.

**Comment by MAX-786** [comment]
> You won't believe this, but https://github.com/executablebooks/sphinx-book-theme/pull/666 that partially includes this feature. I think our implementation is better. 😁

Woohoo! Thanks to you😌🥂 Theirs look bulky😂. 

> 1. On the Search page, the keyboard shortcut needs to be moved a little to the left, so that it does not touch the "search" button.

On my screen, it looked okay, I also checked for its responsiveness as guided by you. But I'll look from my side why this is happening and brb 👍🏻
Screenshot from preview opened on my OS (Win64)
Screenshot_20230105_183130

**Comment by MAX-786** [comment]
> 1. On the Search page, the keyboard shortcut needs to be moved a little to the left, so that it does not touch the "search" button.

Please can you check if it's fixed or not?

except for this, rest stuff is Done!

**Comment by stevepiercy** [comment]
It needs to go about 3px left.

Screen Shot 2023-01-05 at 12 17 09 PM

**Comment by stevepiercy** [comment]
A wee bit more. It needs to clear the button on focus, too.

Screen Shot 2023-01-05 at 1 01 27 PM

**Comment by MAX-786** [comment]
sorry again .... I am visually impaired (no Mac OS) to see actual distance😖

**Comment by stevepiercy** [comment]
Looks good now.

Screen Shot 2023-01-05 at 1 25 15 PM
Screen Shot 2023-01-05 at 1 25 11 PM

I approve, but because this is a large change for search, and @ksuess has been the lead on it, I want to wait for her to have the chance to give a review as a courtesy. Let's wait a week.

**Review by stevepiercy — DISMISSED** [review]
Excellent work!

**Comment by MAX-786** [comment]
Sure! Thanks 😊

**Comment by ksuess** [comment]
Great job you both! I really appreciate this enhancement in UX having a shortcut available!

I think the shortcut label is a hint not a button to press, so I personally would prefer something that reflects this fact. With fresh new years energy, I played a bit with the search forms. So shame on me, that I do present a PR with my suggestions. Feel free to pick what we can aggree upon.

My comments
- Keys have their own tag: `<kbd>`. We can use them to indicate, that keys can lead the user to the search input field.
- Ellipsis is for shorten text. So I would prefer "Search the docs" instead of "Search the docs…"
- Bootstrap (in our case Bootsrap 4 https://getbootstrap.com/docs/4.0/components/input-group/) helps designing forms a lot. See  class="input-group-prepend"

**Comment by MAX-786** [comment]
Thanks a lot for your time! Really appreciated 😊

**Comment by MAX-786** [comment]
@stevepiercy , @ksuess  sorry for late response, taking things from #1421 and following [Steve's comment](https://github.com/plone/documentation/pull/1421#issuecomment-1377087331) (which concluded all the stuff we need for search-field) I have implemented all those things, but before I do implement the search icon and reset button I have some doubts:
- We decided to remove `reset` button but do we need it's functionality since input of type=search gives us little cross which only clear the input value and didn't have any affect on filter. So, for now I have commented it out but is it enough or I should do what ksuess said:
  > A nice to have would be the elimination of the 'reset' by a little cross appearing for a filled search field. ;-)

- Now we have placeholder='Search' ,I need help on how I can use FontAwesome icons or bootstrap icons bundled with pyData sphinx theme in `search-field's placeholder`.

**Comment by ksuess** [comment]
>     * Now we have placeholder='Search' ,I need help on how I can use FontAwesome icons or bootstrap icons bundled with pyData sphinx theme in `search-field's placeholder`.
Please have a look at Sphinx book theme site https://sphinx-book-theme.readthedocs.io/en/stable/. They include a `<i class="icon fas fa-search"></i>` after the input field.
And an 'align-items-center' class for the flex container helps poking the icon on the center line.

File: docs/_templates/search-field.html
Comment (ksuess) [inline docs/_templates/search-field.html]: The outline is nicer with the dedicated bootstrap class.
```suggestion
    <div class="input-group-append">
      <span class="input-group-text" id="shortcut">
        <kbd id="search-shortcut">⌘</kbd>&nbsp;<kbd>K</kbd>
      </span>
    </div>
```
and move after the input field (and the to be added search icon)

**Review by ksuess — CHANGES_REQUESTED** [review]

**Comment by ksuess** [comment]
>     * We decided to remove `reset` button but do we need it's functionality since input of type=search gives us little cross which only clear the input value and didn't have any affect on filter. So, for now I have commented it out but is it enough or I should do what ksuess said:
>       > A nice to have would be the elimination of the 'reset' by a little cross appearing for a filled search field. ;-)
Good catch! Yes, I think a reset should empty the input field and select option "all documentation".
Maybe you could come up with a form action for this?
For the layout: We could replace the "reset" with a cross icon. And move the cross icon to the left of "search" button.

**Comment by MAX-786** [comment]
This failing test is due to missing `CHANGELOG.md`, anything I can do to resolve it? Please help :)

**Comment by ksuess** [comment]
**Sidebar**: @MAX-786, the search field in the sidebar looks pretty good. But would you please be so kind and take a look at https://getbootstrap.com/docs/4.0/components/input-group/
It explains how to use bootstrap classes to glue elements to the input field without the need to fiddle with padding, stretching, etc in CSS.
Bootstrap 4 is part of the theme, so we can use it.
Keep an eye on the order of html nodes: nodes to be appended after the main node.

Bootstrap also helps with the alignment to the center line: 
`<div class="input-group align-items-center">`
See https://getbootstrap.com/docs/4.0/utilities/flex/#align-items

And I think we said to hide the short cut label on focus.
My first approach would be one with jquery .focus() and .hide().
https://api.jquery.com/focus/

**Search page**:
The clear/reset needs to clear also the search results.
I am aware that the current default text 'Your search did not match any documents. Please make sure that …' is not the yellow of an egg; we do step by step.

Please move the shortcut layout to the end of the input field.

If there is no other reason to use svg code, please use icons like in sidebar navigation:
`<i class="fas fa-chevron-down" />`
Font Awesome is included. No need to add it in [conf.py](http://conf.py/).

**Comment by stevepiercy** [comment]
> This failing test is due to missing `CHANGELOG.md`, anything I can do to resolve it? Please help :)

This can be safely ignored. Volto docs pulls in `CHANGELOG.md` from its repo, and Docs pulls in Volto docs. Thus `CHANGELOG.md` has two different relative paths, depending on the context of whether you build docs in Documentation or Volto. See https://github.com/plone/volto/pull/3994/files#diff-8ea15a25d044c7930c6e3f0f521299f7b93d2d8196a08ac0e1c4898f73869d5fR12-R22

However, if this becomes annoying, I can make the `CHANGELOG.md` not get included in Volto's local build, and only included in Documentation's build. That might be a more sane approach. @ksuess you hit this, too. Thoughts?

**Comment by MAX-786** [comment]
Sorry for delayed response, I am bit sick and laptop is not allowed at hospital😑, I'll look into this as soon as I get home.👍

**Comment by ksuess** [comment]
> Sorry for delayed response, I am bit sick and laptop is not allowed at hospital😑, I'll look into this as soon as I get home.👍

I am sorry to hear this, get well soon!
The PR is in no hurry, take your time.

**Comment by MAX-786** [comment]
Thanks for the wait! :)
Ready for review. @ksuess 

> 5. For now, I am OK with making the search form not visible in the sidebar on the Search page, even though it takes up space. I think it might be jarring to the user if the navigation jumps up and down between searching and not searching. Navigation should not move around.

@stevepiercy, that empty space on sidebar (when we're on Search page) was looking bad and after trying it myself, navigation's jumping (up/down) is not causing that much jarring to the user. Please can you try it on preview and provide your feedback, if it isn't looking I'll revert it back. Thanks! :)

**Comment by MAX-786** [comment]
> I am aware that the current default text 'Your search did not match any documents. Please make sure that …' is not the yellow of an egg; we do step by step.

@ksuess, after a little bit of investigating I can think of 2 approaches to handle this:

1. make changes directly in [searchtools.js](https://github.dev/plone/documentation/blob/79c2d954cb84187d63833945689e7a4e7ff720a3/docs/_static/searchtools.js#L346) to show different status conditionally (not exactly sure how).😅
2. basic approach would be that we let `searchtools.js` render the status (search summary) and then in our `search-shortcut.js` inside the `onReset()` function we will replace the text.

Thanks! :) your feedback is most welcome.

**Comment by MAX-786** [comment]
One more thing!

Do we need visual indications when the user is on mobile phone? IG not, right?

**Comment by stevepiercy** [comment]
> Do we need visual indications when the user is on mobile phone?

Yes, we do. I have one of these for my phone, and I have seen people at PloneConf using tablets with similar keyboards.

https://smile.amazon.com/gp/product/B08PL67TRR/

**Comment by MAX-786** [comment]
>  Yes, we do. I have one of these for my phone, and I have seen people at PloneConf using tablets with similar keyboards.

Ohh that looks so cool.. 😯👍🏻 ...

**Review by ksuess — CHANGES_REQUESTED** [review]
This looks pretty good. 
Just one minor thing: The short cut label is somehow bumping when I reset the form or select an option. I think this could be reduced by show/hide the shortcut label with CSS instead of JS.
May I ask you to try replacing js show/hide with something like:
```
#search-form:focus-within #shortcut-page,
#search-form:focus-within #shortcut {
  display: none;
}
```

**Comment by ksuess** [comment]
> > I am aware that the current default text 'Your search did not match any documents. Please make sure that …' is not the yellow of an egg; we do step by step.
> 
> @ksuess, after a little bit of investigating I can think of 2 approaches to handle this:
> 
>     1. make changes directly in [searchtools.js](https://github.dev/plone/documentation/blob/79c2d954cb84187d63833945689e7a4e7ff720a3/docs/_static/searchtools.js#L346) to show different status conditionally (not exactly sure how).😅
> 
>     2. basic approach would be that we let `searchtools.js` render the status (search summary) and then in our `search-shortcut.js` inside the `onReset()` function we will replace the text.
> 
> 
> Thanks! :) your feedback is most welcome.

My preference would be 1. setting search status in `searchtools.js`
If you aggree, a distinction on `query`:
If no `query` then text 'No query, no results' (that's what MDN shows)
if `query`, than the text like it is.
No translation needed, we only have english language.

**Comment by ksuess** [comment]
> One more thing!
> 
> Do we need visual indications when the user is on mobile phone? IG not, right?

Please help me, on which visual indications do you think?

**Comment by MAX-786** [comment]
> Please help me, on which visual indications do you think?

yes, I meant for shortcut label (visual indication) but as Steve said ppl do use keyboards with tablets and phones, so hiding shortcut label on smaller devices isn't needed.

**Comment by ksuess** [comment]
Thanks a lot for the fixes! 
I just adjusted a border-radius.
I think it is good to merge.
If you still have energy, you may want to have look at the border-radius of the search field in the sidebar. 
Maybe its possible to rearrange the input group. But it still needs some CSS for the input group :focus-within to have rounded border on the right.

**Comment by MAX-786** [comment]
Thanks a lot for your guidance :) and sorry I may have asked too many doubts 😅

> If you still have energy, you may want to have look at the border-radius of the search field in the sidebar.

oooo you may already know that students are always full of energy😂. Surely I'll look into it.

**Review by ksuess — DISMISSED** [review]
That was fast! Awesome!

Good job. I would say ready to merge. What do you think, @stevepiercy?

**Comment by stevepiercy** [comment]
Please give me a couple of days to review. I'm under a deadline for $WORK. Thank you!

**Comment by ksuess** [comment]
I've read again @stevepiercy 's suggestion we agreed upon: https://github.com/plone/documentation/pull/1421#issuecomment-1377087331
Sorry for not being attentive.
I think according what we agreed on, there is the placeholder to be removed. And also the glass on the left to be removed in favor of a search button with a glass instead of text 'search'.
So something like 
Search_—_Plone_Documentation_v6_0_new
Which would look much cleaner.
If I am correct, @stevepiercy, then: @MAX-786 would you be so kind and have another look, that would be great.
 
The reason for not removing the clear button is that not all browsers do implicitly show a little cross for input fields of type 'search'.

**Comment by MAX-786** [comment]
> I think according what we agreed on, there is the placeholder to be removed. And also the glass on the left to be removed in favor of a search button with a glass instead of text 'search'.

@ksuess Yes, I'm in favour of moving glass icon inside the button but IMHO we should let placeholder ('Search') there since you first focus on the input field and not on the button ,so we need placeholder for visual indication for user.

**Comment by MAX-786** [comment]
@stevepiercy and @ksuess, just a polite reminder :). This PR can have some love. Thanks!

**Comment by stevepiercy** [comment]
I haven't forgotten. $WORK has taken priority the past month. A review is in my queue, and this one takes more thought than usual. Thanks for your patience.

**Review by stevepiercy — APPROVED** [review]
I ❤️ it!

Let's get @ksuess's latest opinion, and if she approves, we can merge.

Thank you!

**Comment by stevepiercy** [comment]
BTW, the [failing build](https://github.com/plone/documentation/actions/runs/4084515935/jobs/7041306623) is due to ongoing slow response times from `archive.org`. I might need to start excluding it from `make linkcheckbroken`.

**Comment by MAX-786** [comment]
> I haven't forgotten. $WORK has taken priority the past month. A review is in my queue, and this one takes more thought than usual. Thanks for your patience.

Thanks for reviewing! :)

> BTW, the [failing build](https://github.com/plone/documentation/actions/runs/4084515935/jobs/7041306623) is due to ongoing slow response times from archive.org. I might need to start excluding it from make linkcheckbroken.

oh that's why re-running that job passes the test.

**Review by ksuess — APPROVED** [review]

**Comment by ksuess** [comment]
Good job! And an awesome enhancement, thank you @MAX-786!

**Comment by MAX-786** [comment]
Thanks a lot, @stevepiercy and @ksuess, for your guidance😊. 

I am willing to participate in the upcoming GSoC and wanted to know if there is any project in which I can help you and if you may mentor that project in GSoC.
I have been thinking about the Nuclia search project, is there any lead on it?

**Comment by stevepiercy** [comment]
@MAX-786 the GSoC projects are nominated by the mentors, then discussed and selected by them. I nominated Nuclia search. Now we wait for the discussion.

Meanwhile, you can follow this post: https://community.plone.org/t/recommendations-for-gsoc-aspirants/16301

**Comment by polyester** [comment]
To be very precise: the mentors nominate projects, but students can also write proposals*. However, the chances of success of student-nominated projects increase astronomically if they are more or less aligned with mentor proposals. So, early engagement at community.plone.org is the way to go.

*(those are GSOC rules, over which we have no direct control)

**Comment by MAX-786** [comment]
> @MAX-786 the GSoC projects are nominated by the mentors, then discussed and selected by them. I nominated Nuclia search. Now we wait for the discussion.
>  
> Meanwhile, you can follow this post: https://community.plone.org/t/recommendations-for-gsoc-aspirants/16301

Yep, I read the whole thread. and read your proposal about Nuclia Search which we also discussed in discord too. And I'm keeping an eye on it.

**Pull request merged**
